src/sinks/util/auto_concurrency/controller.rs

```diff
@@ -0,0 +1,243 @@
+            // concurrency limit.
+            if inner.current_limit > 1 && (inner.had_back_pressure || rtt >= avg + threshold) {
+                // Decrease (multiplicative) the current concurrency limit
+                let to_forget = inner.current_limit / 2;
+                self.semaphore.forget_permits(to_forget);
+                inner.current_limit -= to_forget;
+            }
+            // Normal quick responses trigger an increase in the
+            // concurrency limit. Note that we only check this if we had
+            // requests to go beyond the current limit to prevent
+            // increasing the limit beyond what we have evidence for.
+            else if inner.current_limit < self.max
+                && inner.reached_limit
+                && !inner.had_back_pressure
+                && rtt <= avg
+            {
+                // Increase (additive) the current concurrency limit
+                self.semaphore.add_permits(1);
+                inner.current_limit += 1;
+            }
+            emit!(AutoConcurrencyLimit {
+                concurrency: inner.current_limit as u64,
+            });
+
+            // Reset values for next interval
+            let new_avg = inner.past_rtt.update(rtt);
+            inner.next_update = now + Duration::from_secs_f64(new_avg);
+            inner.current_rtt.reset();
+            inner.had_back_pressure = false;
+            inner.reached_limit = false;
+        }
+    }
+}
+
+impl<L> Controller<L>
+where
+    L: RetryLogic,
+{
+    pub(super) fn adjust_to_response(
+        &self,
+        start: Instant,
+        response: &Result<L::Response, crate::Error>,
+    ) {
+        let is_back_pressure = match response {
+            Ok(_) => false,
+            Err(err) => {
+                if let Some(err) = err.downcast_ref::<L::Error>() {
+                    self.logic.is_retriable_error(err)
+                } else if err.downcast_ref::<Elapsed>().is_some() {
+                    true
+                } else {
+                    panic!("Unhandled error response! {:?}", err)
+                }
+            }
+        };
+        self.adjust_to_back_pressure(start, is_back_pressure)
+    }
+}
+
+/// Exponentially Weighted Moving Average
+#[derive(Clone, Copy, Debug, Default)]
+struct EWMA {
+    average: f64,
+}
+
+impl EWMA {
+    fn average(&self) -> f64 {
+        self.average
+    }
+
+    /// Update and return the current average
+    fn update(&mut self, point: f64) -> f64 {
+        self.average = match self.average {
+            avg if avg == 0.0 => point,
+            avg => point * EWMA_ALPHA + avg * (1.0 - EWMA_ALPHA),
+        };
+        self.average
+    }
+}
+
+/// Simple unweighted arithmetic mean
+#[derive(Clone, Copy, Debug, Default)]
+struct Mean {
+    sum: f64,
+    count: usize,
+}
+
+impl Mean {
+    /// Update and return the current average
+    fn update(&mut self, point: f64) -> f64 {
+        self.sum += point;
+        self.count += 1;
```

Comment:
Do you think we need to worry about overflow for either of these?

Reply:
I doubt it. `sum` is a `f64`, and it would be pretty astonishing for that to overflow in human time scales. Similarly, `count` is `usize`, which will take a rather impossibly large number of packets to overflow.